sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp

```diff
@@ -0,0 +1,317 @@
+    const auto thread_y = idx / Dim;
+
+    if constexpr (Layout == matrix_layout::col_major) {
+      for (int i = 0; i < 4; ++i) {
+        const int c_idx = thread_x * K + i + thread_y * 4;
+        res.data[i] = src[c_idx];
+      }
+    } else if constexpr (Layout == matrix_layout::row_major) {
+      for (int i = 0; i < 4; ++i) {
+        const int r_idx = thread_x + i * NumCols + thread_y * NumCols * 4;
+        res.data[i] = src[r_idx];
+      }
+    }
+  }
+}
+
+template <typename Group, matrix_layout Layout, typename T, size_t NumRows,
+          size_t NumCols, access::address_space Space,
+          access::decorated IsDecorated>
+void store_layoutT(joint_matrix_hip<T, matrix_use::accumulator, NumRows,
+                                    NumCols, matrix_layout::dynamic> &src,
+                   multi_ptr<T, Space, IsDecorated> dst, size_t stride,
+                   Group &sg) {
+  std::ignore = stride;
+  const auto idx = sg.get_group_linear_id() * sg.get_local_range()[0] +
+                   sg.get_local_linear_id();
+
+  if constexpr (std::is_same_v<T, double>) {
+    if constexpr (NumRows == 16 && NumCols == 16) {
+      const auto thread_x = idx % NumCols;
+      const auto thread_y = idx / NumCols;
+      for (int i = 0; i < 4; ++i) {
+        dst[thread_x + i * 4 * NumCols + thread_y * NumCols] = src.data[i];
+      }
+    }
+  } else if constexpr (std::is_same_v<T, float> || std::is_same_v<T, int32_t>) {
+    if constexpr (NumRows == 16 && NumCols == 16) {
+      const auto thread_x = idx % NumCols;
+      const auto thread_y = idx / NumCols;
+      constexpr int K = 16;
+
+      for (int i = 0; i < 4; ++i) {
+        const int d_idx = thread_x + i * K + thread_y * 4 * K;
+        dst[d_idx] = src.data[i];
+      }
+    } else if constexpr (NumRows == 32 && NumCols == 32) {
+      const auto thread_x = idx % NumCols;
+      const auto thread_y = idx / NumCols;
+      constexpr int K = 8;
+
+      for (int j = 0; j < 4; ++j) {
+        for (int i = 0; i < 4; ++i) {
+          const int d_idx =
+              thread_x + i * NumCols + thread_y * 4 * NumCols + j * 8 * NumCols;
+          dst[d_idx] = src.data[i + 4 * j];
+        }
+      }
+    }
+  } else {
+    static_assert(false && "Invalid dimenstions!");
+  }
+}
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          access::address_space Space, access::decorated IsDecorated>
+void joint_matrix_store_hip(
+    joint_matrix_hip<T, matrix_use::accumulator, NumRows, NumCols,
+                     matrix_layout::dynamic> &src,
+    multi_ptr<T, Space, IsDecorated> dst, size_t stride, matrix_layout Layout,
+    Group &sg) {
+  switch (Layout) {
+  case matrix_layout::row_major:
+    store_layoutT<Group, matrix_layout::row_major>(src, dst, stride, sg);
+    break;
+  case matrix_layout::col_major:
+    store_layoutT<Group, matrix_layout::col_major>(src, dst, stride, sg);
+    break;
+  default:
+    assert(false && "Invalid layout specified!");
```

Comment:
I think only row_major or col_major layouts can be specified? Maybe I am wrong, but if this is the case such asserts aren't required.